doc: Adding nrf54l crypto doc to dev guide (#17461)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/nrf54l/cryptography.rst

.. _ug_nrf54l_cryptography:

nRF54L Series cryptography
##########################

.. contents::
   :local:
   :depth: 2

The cryptographic peripherals of the nRF54L Series are supported through a set of standard PSA Crypto APIs, with some additional vendor specific extensions.

The nRF Security library offers a set of :ref:`nrf_security_drivers`.

> On the nRF54L devices, it is the CRACEN driver (nrf_cracen) that provides entropy and hardware-accelerated cryptography using the CRACEN (Crypto Accelerator Engine) peripheral.

Reviewer: Oberon SW crypto driver is also supported on the nRF54L right?

Some KMU slots are reserved for current and future |NCS| use cases.

.. list-table:: List of reserved KMU slots
   :widths: auto
   :header-rows: 1

   * - | Reserved KMU slots
       | (range inclusive)
     - NCS usage
     - Description
   * - 180-182
     - Reserved
     - --
   * - 183-185
     - IKG seed
     - 384-bit random seed to generate keys using the CRACEN IKG
   * - 186-225
     - Reserved
     - --
   * - 226-227
     - UROT_PUBKEY_0
     - | Revokable MCUboot firmware image key, generation 0.
       | ED25519 public key.
   * - 228-229
     - UROT_PUBKEY_1
     - | Revokable MCUboot firmware image key, generation 1.
       | ED25519 public key.
   * - 230-231
     - UROT_PUBKEY_2
     - | Revokable MCUboot firmware image key, generation 2.
       | ED25519 public key.
   * - 232-241
     - Reserved
     - --
   * - 242-243
     - BL_PUBKEY_0
     - | Revokable NSIB firmware image key, generation 0.
       | ED25519 public key.
   * - 244-245
     - BL_PUBKEY_1
     - | Revokable NSIB firmware image key, generation 1.
       | ED25519 public key.
   * - 246-247
     - BL_PUBKEY_2
     - | Revokable NSIB firmware image key, generation 2.

>        | ED25519 public key.

Reviewer 2: Not an expert in naming stuff, but MCUboot can serve as NSIB as well, so shouldn't the naming be rather Non-mutable and  mutable bootloader or something like this?

Author: Changed to upgradable and immutable.

   * - 248-255
     - Reserved
     - --

> Programming model for referencing keys

Reviewer: In this section, we should mention https://arm-software.github.io/psa-api/ or https://arm-software.github.io/psa-api/crypto/1.2/, so users know which psa functions we refer to in the table.

|                                      |                                                                             | For correct values for kmu_slot_nr, see :ref:`ug_nrf54l_crypto_kmu_slots`.                                                    |
+--------------------------------------+-----------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------+
| key_usage (psa_set_key_usage_flags)  | Standard PSA Crypto key usage flags                                         | PSA_KEY_USAGE_EXPORT and PSA_KEY_USAGE_COPY are not allowed for keys with the usage scheme CRACEN_KEY_USAGE_SCHEME_PROTECTED. |
+--------------------------------------+-----------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------+

.. _ug_nrf54l_crypto_kmu_key_usage_schemes:

> KMU key usage schemes

Reviewer: Are there any of these users are not supposed to use? If so we should write that.
Is there one of these we recommend to use by default? I think we should recommend one so that users who really don't know end up picking something good and not just pick one at random

Reviewer: Maybe we could even have a "typical use-case" column?

Reviewer: If we list which schemes support which keys, and note that SEED is usually not relevant for users, this will be good.

     - 1
     - Yes
     - Yes
     - Yes
   * - AES 192-bit and 256-bit keys
     - | key_type: PSA_KEY_TYPE_AES
       | key_bits: 192
       | key_bits: 256
     - 2
     - Yes
     - Yes
     - Yes
   * - ChaCha20-Poly1305
     - key_type: PSA_KEY_TYPE_CHACHA20
     - 2
     - No
     - Yes
     - Yes
   * - ECC secp256r1 key pair
     - | key_type: PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1)
       | key_bits: 256
     - 2
     - No
     - Yes
     - Yes
   * - ECC secp256r1 public key
     - | key_type: PSA_KEY_TYPE_ECC_PUBLIC_KEY(PSA_ECC_FAMILY_SECP_R1)
       | key_bits: 256
     - 4
     - No
     - Yes
     - Yes
   * - Ed25519 key pair
     - | key_type: PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_TWISTED_EDWARDS)
       | key_bits: 255
     - 2
     - No
     - Yes
     - Yes
   * - ED25519 public key
     - | key_type: PSA_KEY_TYPE_ECC_PUBLIC_KEY(PSA_ECC_FAMILY_TWISTED_EDWARDS)
       | key bits: 255
     - 2
     - No
     - Yes
     - Yes
.. [1] Keys with the Encrypted usage scheme (CRACEN_KMU_KEY_USAGE_SCHEME_ENCRYPTED) will require two additional KMU slots to store the nonce and the authentication tag.

Storing keys in KMU
-------------------

> Applications can store keys in KMU slots using the standard PSA cryptographic key management operations psa_import_key, psa_generate_key or psa_copy_key.

Reviewer: I know it is another topic, but we should really integrate [PSA Crypto API]( https://arm-software.github.io/psa-api/crypto/1.2/) docs into our docs soon, so we can link these API calls.

.. note::
   If there is a power failure when provisioning a key with persistence PSA_KEY_PERSISTENCE_READ_ONLY or CRACEN_KEY_PERSISTENCE_REVOKABLE, it might not be possible to recover the key slot.
   Provisioning of read-only keys should be restricted to controlled environments (production).

The following special error codes are available:

* PSA_ERROR_ALREADY_EXIST: One of the required key slots has already been provisioned
* PSA_ERROR_NOT_SUPPORTED: Unsupported key type

Removing or revoking keys from the KMU
--------------------------------------

Keys are deleted or revoked using the psa_destroy_key function.
Calling the psa_destroy_key function on keys that have the persistence CRACEN_KEY_PERSISTENCE_REVOKABLE, will mark the associated KMU slots as revoked, preventing the slots from being re-used for new keys.

Using KMU keys
--------------

Keys stored in the KMU can be used in standard PSA cryptographic operations for encryption, decryption, signing a hash or a message, and verifying a hash or a message, given that the corresponding PSA_KEY_USAGE_* flags are set.

> KMU push operations

Reviewer: Should this be a heading?

Keys with the usage schemes CRACEN_KMU_KEY_USAGE_SCHEME_PROTECTED and CRACEN_KMU_KEY_USAGE_SCHEME_SEED can push data from the slots directly to CRACEN registers that are not accessible by the CPU.
These usage schemes are supported for pushing symmetric keys used for cipher operations, and for seeds for use by the CRACEN IKG.

> Keys with the usage scheme CRACEN_KMU_KEY_USAGE_SCHEME_RAW, such as symmetric keys, must be temporarily pushed to a RAM location, before CRACEN can load them into the asymmetric engine before running operations like sign or verify.

Reviewer: >  must be temporarily pushed to a RAM location

"Must be"? Will the user have to do this or is it done automatically by our driver? If the latter, maybe "will be" instead?

> Key slots that are tagged with CRACEN_KMU_KEY_USAGE_SCHEME_ENCRYPTED also have to be decrypted to a temporary push location in RAM before they are used by CRACEN.

Reviewer: Do users have to decrypt these or is that done by the driver? If done by the driver, this should be more clear.
If done by the user, we should say how they should do it.

Reviewer 3: Mentioning the driver here makes sense indeed. And the same goes for your previous (good) point about the keys having to be pushed to ram.

When the application is built with TF-M, this temporary push location is protected inside the secure processing environment to avoid exposing the key material to the non-secure application.
If TF-M is not used, the keys are pushed to a reserved RAM area at location 0x20000000-0x20000064 (kmu_push_area).

> The following special error codes are available:

Reviewer: The PSA Crypto API list error codes per function. Do these error codes come from one specific function or from any PSA Crypto API function?

* PSA_ERROR_INVALID_HANDLE: Attempting an operation is attempted on an empty KMU slot
* PSA_ERROR_NOT_PERMITTED: Trying an operation on a revoked key
* PSA_ERROR_HARDWARE_FAILURE: Key slot has invalid data
* PSA_ERROR_CORRUPTION_DETECTED: Key slot has invalid data

> CRACEN IKG

Reviewer: Maybe we should draw the parallell to HW unique keys here? Just so users who know the HW Unique Keys easily can re-understand the use-case of IKG

The derived keys are also not accessible by the CPU, but they can be used for cryptographic operations by CRACEN.

These keys are also accessed using normal PSA Crypto APIs, and are referenced by special built-in key IDs.

+-----------------+---------------------------------+---------------------------------------------------+
| Key type        | Key ID                          | Description                                       |
+=================+=================================+===================================================+
| ECC secp256r1   | CRACEN_BUILTIN_IDENTITY_KEY_ID  | Used for signing/verification.                    |
+-----------------+---------------------------------+---------------------------------------------------+
| AES 256-bit     | CRACEN_BUILTIN_MKEK_ID          | Used for encryption/decryption or key derivation. |
+-----------------+---------------------------------+---------------------------------------------------+
| AES 256-bit     | CRACEN_BUILTIN_MEXT_ID          | Used for encryption/decryption or key derivation. |
+-----------------+---------------------------------+---------------------------------------------------+

> 

Reviewer 4: Maybe also worth mentioning to the reader: 

`CRACEN_BUILTIN_MKEK_ID` and `CRACEN_BUILTIN_MEXT_ID` are master keys and they cannot be used directly. The key derivation algorithm `PSA_ALG_SP800_108_COUNTER_CMAC` should be used to derive keys first. https://arm-software.github.io/psa-api/crypto/1.2/api/ops/kdf.html#c.PSA_ALG_SP800_108_COUNTER_CMAC.

Reviewer 4: Reading this section again, it seems to just be mentioned to the user for informational purposes and end-users aren't actually meant to use these keys in their code. If that is the case, we don't need to add this extra piece of info.

Reviewer 4: We'll leave this section as it is, although with a note that this is not really meant for end-users to be using directly.

The keys are not exportable, except for the public key associated with the asymmetric key.

> Configuration

Reviewer: No Kconfig options needed to enable IKG?

Reviewer 3: It might makes sense to clarify though that the IKG keys are derived using the seed. The seed is generated automatically during the first boot of the device and it is stored in the KMU.  So erasing the device will definitely change the seed and thus the IKG keys.

Reviewer: From the seed generation explanation, I found [CONFIG_CRACEN_LOAD_KMU_SEED](https://docs.nordicsemi.com/bundle/ncs-latest/page/kconfig/index.html#CONFIG_CRACEN_LOAD_KMU_SEED). 
We should mention this, as it seems it will both load the seed, and that this config will cause the seed to be generated at first boot (or if the seed does not exist for some other reason).

Reviewer: Also, I think that in the future, we will have some provisioning step for the nRF54L15, where the seed will be generated. At that point we should remember to add a description here that the seed is meant to be generated by such provisioning

=============

To enable KMU support, do the following:

.. code-block:: console

   CONFIG_NRF_SECURITY=y
   CONFIG_MBEDTLS_PSA_CRYPTO_C=y
   CONFIG_PSA_CRYPTO_DRIVER_CRACEN=y

Enabling support for encrypted key scheme (CRACEN_KMU_KEY_USAGE_SCHEME_ENCRYPTED), should look as follows:

.. code-block:: console

   CONFIG_PSA_WANT_ALG_GCM=y
   CONFIG_PSA_WANT_KEY_TYPE_AES=y
   CONFIG_PSA_WANT_AES_KEY_SIZE_256=y

>    CONFIG_PSA_WANT_ALG_SP800_108_COUNTER_CMAC=y

Reviewer 4: I checked, we need these two configs as well to enable KMU support

CONFIG_PSA_WANT_ALG_CMAC=y
CONFIG_PSA_WANT_ALG_ECB_NO_PADDING=y

Reviewer: At which point should we we have some umbrella config so users don't have to add all these to "clutter" their project config to use the KMU?

Reviewer 4: I agree, would be nice to have one config to enable KMU support. Wonder if this is the right place to add a task for this https://nordicsemi.atlassian.net/browse/NCSDK-7747

|                                              |                                                                             | See :ref:`ug_nrf54l_crypto_kmu_supported_key_types` for overview of the supported key types for each driver.                              |
+----------------------------------------------+-----------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
| ``key lifetime`` (``psa_set_key_lifetime``)  | ``PSA_KEY_LIFETIME_FROM_PERSISTENCE_AND_LOCATION(persistence, location)``   | ``CRACEN_KEY_PERSISTENCE_REVOKABLE`` is a custom persistence mode, which will revoke the key slots when the key is destroyed.             |
|                                              | where persistence is set to one of the following:                           |                                                                                                                                           |

> |                                              | ``PSA_KEY_PERSISTENCE_DEFAULT``,                                            |                                                                                                                                           |

Reviewer 2: What does default do?

Reviewer: From [here](https://github.com/hellesvik-nordic/sdk-nrf/blob/e6eec800a3a7cd53a113ab557bc73ff8523540cd/subsys/nrf_security/src/drivers/cracen/cracenpsa/src/kmu.c#L663-L664) we can see that it is LIB_KMU_REV_POLICY_ROTATING. 
The behaviour of this is explained under the OPS -> KMU -> 7.7.3.2 Operations.

Author: Added a bit more info on default and read-only persistence.

   * - Ed25519 key pair
     - | ``key_type``: ``PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_TWISTED_EDWARDS)``
       | ``key_bits``: 255
     - 2
     - No
     - Yes
     - Yes
   * - ED25519 public key
     - | ``key_type``: ``PSA_KEY_TYPE_ECC_PUBLIC_KEY(PSA_ECC_FAMILY_TWISTED_EDWARDS)``
       | ``key bits``: 255
     - 2
     - No
     - Yes
     - Yes
.. [1] Keys with the Encrypted usage scheme (``CRACEN_KMU_KEY_USAGE_SCHEME_ENCRYPTED``) will require two additional KMU slots to store the nonce and the authentication tag.

Storing keys in KMU
-------------------

Applications can store keys in KMU slots using the standard PSA cryptographic key management operations ``psa_import_key``, ``psa_generate_key`` or ``psa_copy_key``.
Additionally, the KMU slots can be provisioned using the nRF Util development tool.

.. note::
   If a power failure occurs during provisioning of a key with persistence ``PSA_KEY_PERSISTENCE_READ_ONLY`` or ``CRACEN_KEY_PERSISTENCE_REVOKABLE``, it might not be possible to recover the key slot.
   Provisioning of read-only keys should be restricted to controlled environments (production environments).

The following special error codes are available:

> * ``PSA_ERROR_ALREADY_EXIST``: One of the required key slots has already been provisioned.

Author: After discussing a bit with the developers, we will leave this information out for the time being.

.. _ug_nrf54l_cryptography:

nRF54L Series cryptography
##########################

.. contents::
   :local:
   :depth: 2

The cryptographic peripherals of the nRF54L Series are supported through a set of standard PSA Crypto APIs, with some additional vendor specific extensions.

The nRF Security library offers a set of :ref:`nrf_security_drivers`.

> On the nRF54L devices, in addition to the nrf_oberon driver covering the software-based cryptography implementations, the CRACEN driver (nrf_cracen) provides entropy and hardware-accelerated cryptography using the CRACEN (Crypto Accelerator Engine) peripheral.

Reviewer 5: Should the nRF Security library page link to this page and mention that there is a nRF54L-specific implementation?

Author: Link added in PR https://github.com/nrfconnect/sdk-nrf/pull/17804, a draft PR with work in progress.

.. _ug_nrf54l_cryptography:

> nRF54L Series cryptography

Reviewer 6: @mia-ko Isn't this applicable also to 54H?

As I believe it is in scope of 2.8 to have support for crypto PSA on both, are we planning to duplicate this page in both "working with" sections or will this be moved somewhere else (like in app dev)? @annwoj  FYI

Author: Crypto for 54H Series is not the same as for 54L Series.

> nRF54L Series cryptography

Reviewer 5: Please add link to this  page in the table on https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/security.rst

Not sure if this should be a separate row under Cryptographic operations or an addition to that row.

Author: Link added in PR https://github.com/nrfconnect/sdk-nrf/pull/17804, a draft PR with work in progress.

Not sure if it should be a separate row either...
I added it along with the Crypto operations entry for the moment, what do you think?
If we add similar content for the nRF54H Series devices in time for the release, we should have own entry... Product-specific cryptography operations... or something like that.

These usage schemes are supported for pushing symmetric keys used for cipher operations, and for seeds used by the CRACEN IKG.

Keys with the usage scheme Raw (``CRACEN_KMU_KEY_USAGE_SCHEME_RAW``), such as symmetric keys, are temporarily pushed to a RAM location by the CRACEN driver, and then loaded by CRACEN into the asymmetric engine before running operations like sign or verify.

Key slots with the usage scheme Encrypted (``CRACEN_KMU_KEY_USAGE_SCHEME_ENCRYPTED``) also have to be decrypted to a temporary push location in RAM before they are used by CRACEN, which is handled by the CRACEN driver.

When the application is built with TF-M, this temporary push location is protected inside the secure processing environment to avoid exposing the key material to the non-secure application.
If TF-M is not used, the keys are pushed to a reserved RAM area at location 0x20000000-0x20000064 (``kmu_push_area``).

You might encounter the following KMU-specific error codes when using the KMU keys:

> * PSA_ERROR_INVALID_HANDLE: Attempting an operation on an empty KMU slot.
> * PSA_ERROR_NOT_PERMITTED: Attempting an operation on a revoked key.
> * PSA_ERROR_HARDWARE_FAILURE: The key slot has invalid data.
> * PSA_ERROR_CORRUPTION_DETECTED: The key slot has invalid data.

Reviewer 5: ```suggestion
* ``PSA_ERROR_INVALID_HANDLE``: Attempting an operation on an empty KMU slot.
* ``PSA_ERROR_NOT_PERMITTED``: Attempting an operation on a revoked key.
* ``PSA_ERROR_HARDWARE_FAILURE``: The key slot has invalid data.
* ``PSA_ERROR_CORRUPTION_DETECTED``: The key slot has invalid data.
```
See https://nordicsemi.atlassian.net/wiki/spaces/TECHDOC/pages/120291747/Typographical+conventions - Value

Author: Just a miss I guess. The ones mentioned earlier, under section Storing keys in KMU are ok. Will fix this.

Configuration
*************

The CRACEN peripheral does not require any configuration.
Its operation is ensured by hardware.

To enable KMU support, set the following Kconfig options:

.. code-block:: console

   CONFIG_NRF_SECURITY=y
   CONFIG_MBEDTLS_PSA_CRYPTO_C=y
   CONFIG_PSA_CRYPTO_DRIVER_CRACEN=y

As an example, enabling support for the Encrypted key usage scheme (``CRACEN_KMU_KEY_USAGE_SCHEME_ENCRYPTED``) should look as follows:

.. code-block:: console

   CONFIG_PSA_WANT_KEY_TYPE_AES=y
   CONFIG_PSA_WANT_AES_KEY_SIZE_256=y
   CONFIG_PSA_WANT_ALG_ECB_NO_PADDING=y
   CONFIG_PSA_WANT_ALG_CMAC=y
   CONFIG_PSA_WANT_ALG_SP800_108_COUNTER_CMAC=y

>    CONFIG_PSA_WANT_ALG_GCM=y

Reviewer 5: This is a useful way of providing code, but it's totally ignoring the fact that we have this page that documents how to configure Kconfig: https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/app_dev/config_and_build/kconfig/index.html 

Moreover, using this code-block method makes it impossible to link to the Kconfig options using `:kconfig:option:`, which is something we recommend in guidelines: https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/dev_model_and_contributions/documentation/styleguide.html#linking

Author: Working on it :)

Author: Any better now, @greg-fer? :)

The configuration is enabling the key type (AES) and the key size (256 bits) supported by the Encrypted usage scheme as explained in the :ref:`ug_nrf54l_crypto_kmu_supported_key_types` section, and in addition enabling the following :ref:`cryptographic features <nrf_security_driver_config>` supported by the CRACEN driver:

* Cipher mode: AES ECB (Electronic CodeBook) mode, no padding
* Message Authentication Code (MAC) cipher: cipher-based MAC (CMAC) cipher

> * Key derivation function (KDF) support: SP800-108 CMAC in counter mode

Author: Will add CONFIG_PSA_WANT_ALG_SP800_108_COUNTER_CMAC in the KDF section in the nRF Security lib (https://docs.nordicsemi.com/bundle/ncs-2.7.99-cs1/page/nrf/libraries/security/nrf_security/doc/driver_config.html#key_derivation_function_configurations).

Author: Kconfig added in PR https://github.com/nrfconnect/sdk-nrf/pull/17804, a draft PR with work in progress.

On the nRF54L devices, in addition to the nrf_oberon driver covering the software-based cryptography implementations, the CRACEN driver (nrf_cracen) provides entropy and hardware-accelerated cryptography using the Crypto Accelerator Engine (CRACEN) peripheral.
The CRACEN PSA driver supports the following:

* Executing cryptographic operations using the CRACEN peripheral.
* Importing, using, and revoking keys stored in the Key Management Unit (KMU).
* Pushing symmetric keys directly from the KMU to the CRACEN symmetric engine, without exposing the key material to the CPU.
* Using isolated keys derived from CRACEN's Isolated Key Generator (IKG) for encryption and signing purposes.

The keys that are stored in the KMU or generated by the IKG are referenced using the built-in keys that are key IDs in the range from ``MBEDTLS_PSA_KEY_ID_BUILTIN_MIN`` to and including ``MBEDTLS_PSA_KEY_ID_BUILTIN_MAX``.

KMU and CRACEN peripherals
**************************

The nRF54L Series Crypto Accelerator Engine (CRACEN) and the Key Management Unit (KMU) peripherals, along with the CRACEN PSA driver, are central when ensuring that the assets of an nRF54L device are protected.

> While CRACEN is not accessed by the CPU and typically not directly used by the end-users and their applications, the KMU provides operations to store, use, and remove assets.

Reviewer 7: ```suggestion
While CRACEN is not accessed by the CPU and typically not directly used by the end-users and their applications, the KMU provides operations to store, use, revoke and/or remove assets.
```

Only the KMU is able to push assets to CRACEN's protected RAM and the SEED register.

> The KMU can store cryptographic keys and a 384-bit random seed in key storage slots.

Reviewer 7: ```suggestion
The KMU can store cryptographic keys and 384-bit random seeds for the IKG in key storage slots.
```

The CRACEN PSA driver exposes the KMU operations through standard PSA Crypto API calls, with some vendor-specific extensions.
The following KMU operations are supported:

* Importing/provisioning keys to KMU slots.
* Deleting a key from the KMU, allowing the underlying storage location to be reused.
* Revoking a key from the KMU, preventing reuse of the underlying key slots.
* Directly pushing symmetric keys to CRACEN's engine for symmetric crypto operations, without exposing the key material to the CPU.
* Directly pushing a 384-bit seed to CRACEN's IKG to derive isolated keys.

> * Pushing asymmetric keys to CPU RAM before loading them into CRACEN's engine for asymmetric operations.

Reviewer 7: ```suggestion
* Pushing symmetric or asymmetric keys to CPU RAM before loading them into CRACEN's engine for cryptographic operations.
```

Additionally, the CRACEN PSA driver supports storing encrypted keys in KMU slots, transparently decrypting them to a temporary RAM location before using them in cryptographic operations.

.. _ug_nrf54l_crypto_kmu_slots:

KMU slots
=========

The KMU is partitioned into 256 numbered slots, each capable of storing 128 bits of key material, a 32-bit target push address and 32 bits of metadata.
Storing keys that are larger than 128 bits is supported by using multiple, consecutive slots.

The application can use the KMU slots to store key data for their own purposes.
Some KMU slots are reserved for current and future |NCS| use cases.
The following table gives an overview of the KMU slots and their usage:

.. list-table:: List of reserved KMU slots
   :widths: auto
   :header-rows: 1

   * - | Reserved KMU slots
       | (range inclusive)

>      - | NCS usage

Reviewer 8: ```suggestion
     - | |NCS| usage
```

       | ED25519 public key.
   * - 230-231
     - UROT_PUBKEY_2
     - | Revokable firmware image key for upgradable bootloader, generation 2.
       | ED25519 public key.
   * - 232-241
     - Reserved
     - --
   * - 242-243
     - BL_PUBKEY_0
     - | Revokable firmware image key for immutable bootloader, generation 0.
       | ED25519 public key.
   * - 244-245
     - BL_PUBKEY_1
     - | Revokable firmware image key for immutable bootloader, generation 1.
       | ED25519 public key.
   * - 246-247
     - BL_PUBKEY_2
     - | Revokable firmware image key for immutable bootloader, generation 2.
       | ED25519 public key.
   * - 248-255
     - Reserved
     - --

CRACEN Isolated Key Generator
=============================

CRACEN's Isolated Key Generator (IKG) can derive three keys from a 384-bit seed value.

IKG-generated keys are also called special hardware keys.
These keys are typically not meant for application use.
They are not accessible by any CPU, but they can be used for cryptographic operations by CRACEN, provided directly to the cryptographic engine as a hardware signal.
IKG keys are not retained, and have to be regenerated for every CRACEN power cycle.

> The 384-bit seed value is generated automatically during the first boot of the device using the CRACEN Random Number Generator (RNG), and then stored in the KMU.

Reviewer 7: ```suggestion
The 384-bit seed value is provisioned to the device or generated automatically during the first boot of the device using the CRACEN Random Number Generator (RNG), and then stored in the KMU.
```

Before the keys can be generated, the seed is pushed by the KMU to the SEED register, and then validated.
Generating keys without the valid seed will fail.

IKG keys are also accessed using the standard PSA Crypto APIs, and are referenced by special built-in key IDs.

+-----------------+-------------------------------------+---------------------------------------------------+
| Key type        | Key ID                              | Description                                       |
+=================+=====================================+===================================================+

> | ECC secp256r1   | ``CRACEN_BUILTIN_IDENTITY_KEY_ID``  | Used for signing/verification.                    |

Reviewer 7: This is changing due to PR: https://github.com/nrfconnect/sdk-nrf/pull/17844

The new proposed name is `CRACEN_BUILTIN_XXXX_ID`
